### DISCUSSION TITLE
.NET 8 Blazor Web App Migration - Testing & Breaking Changes

### DISCUSSION THREAD

**Discussion by sbwalker** [oqtane/oqtane.framework#3874] [General]
The migration to the new Blazor Web App approach is now feature complete. It is now a good time for some testing with your custom modules and themes. 

When you do a new installation on 5.1 the default site will be configured to use Global Interactive Rendering with Server Interactivity. This is the same configuration that Oqtane has always utilized by default. In this configuration you should not notice any differences in behavior.

If you would like to test a different configuration, you will need to visit the Site Settings and scroll down to the Hosting Model section:

![image](https://github.com/oqtane/oqtane.framework/assets/4840590/a561e204-8a11-4dc1-8069-ad03b704fd6e)

Render Mode (how components should be rendered)
- Global Interactive Rendering = full interactivity for all components in the entire application (legacy Blazor behavior)
- Static Server Rendering = no interactivity except for specific components which enable it
- Headless API (No Rendering) = no user interface - would only be used in scenarios where you want to use Oqtane as a content database

Interactivity (the approach for making components interactive)
- Server = SignalR (legacy Blazor Server)
- WebAssembly = WebAssembly
- Auto = starts with SignalR and then transitions to WebAssembly automatically

Prerender (whether or not interactive components should be prerendered on  the server before becoming interactive)
- Yes (default)
- No

Hybrid (whether or not you want .NET MAUI client apps to be able to connect to your site, download assemblies, serve content)
- Yes
- No (default)

It should also be noted that although there is not a configuration option, components which use Static Rendering will also have StreamRendering enabled, whereas components which use Interactive Rendering will have StreamRendering disabled (however you can enable this attribute within your own component). This configuration is necessary in order to get Oqtane to render its UI consistently.

From a development perspective, there is only 1 new option that you need to be aware of... a new RenderMode property on the IModuleControl interface. By default this property is set to Interactive - which means that all module components will be interactive by default (to preserve legacy behavior). However this can be modified on a component basis if you want to create a component which is statically rendered. A good example is the Html/Text module's Index component:

```
@code {
    public override string RenderMode => RenderModes.Static;
```

That's all you need to add to your component to make it render statically. 

However you may also need to learn some new techniques if your static component has interactivity requirements. For example, the ModuleMessage component (Oqtane.Client\Modules\Controls\ModuleMessage.razor) has a button to dismiss the message - however this button could not use the standard ```@onclick``` event handler if the component is statically rendered. Instead it needed to use a different approach:

```
            <form method="post" @onsubmit="DismissModal" @formname="@_formname" data-enhance>
                <input type="hidden" name="__RequestVerificationToken" value="@SiteState.AntiForgeryToken" />
                <button type="submit" class="btn-close" aria-label="Close"></button>
            </form>
```

The button is now wrapped in a form. The form needs to include the anti-forgery token. The ```@onsubmit``` event handler which can be leveraged within your component code. 

One additional item which should be noted is that the core UI composition layer has been significantly optimized in 5.1. Many API and database calls have been eliminated which should result in a major performance boost. This should be apparent regardless of whether you are using static or interactive rendering. So please pay attention to the performance characteristics in your testing as well.

**There are not expected to be any breaking changes in your existing modules or themes** (other than the 1 minor item outlined below). So despite the major architectural changes to the framework in this release, the upgrade process should be seamless and third party applications built on Oqtane will get the benefit of the new features without any disruptions.

**Comment by sbwalker** [comment]
So far there is only one "breaking" change which was unavoidable and was caused by the migration to Blazor Web and static rendering. It is not a breaking change which causes a run-time exception... it is a breaking change from a functionality perspective. It is related to direct references to ModuleInstance.

The ModuleInstance component previously contained methods for AddModuleMessage() and ShowProgressIndicator(). These methods were also available in ModuleBase so that developers did not need to reference ModuleInstance directly. In the migration to Blazor Web, there was a need to introduce some additional component abstraction layers between a module component and ModuleInstance. This was needed to enable Interactive or Static render modes as well as StreamRendering for modules... and to pass state across the render mode boundary (from static to interactive). 

Formerly, a reference of the ModuleInstance component was passed using a parameter to the ModuleBase - providing a way to invoke its methods. In the refactoring it is no longer possible to pass a reference of the ModuleInstance component because component references are not serializable - which means they cannot be passed across the render mode boundary. Instead, the AddModuleMessage() and ShowProgressIndicator() methods had to be added on the interactive side of the render mode boundary, in the RenderModeBoundary component. The ModuleBase methods were updated to point to the new RenderModeBoundary component, so they continue to work fine.

However if your module component contained logic such as ModuleInstance.AddModuleMessage("message", MessageType.Info); it will now have a compiler warning when you try to build your code: "AddModuleMessage is deprecated. Use AddModuleMessage in ModuleBase instead.". At run-time, your module will not throw an exception - but it will also not display a message in the UI as intended - the method will be ignored. 

To resolve this issue is very simple... change ModuleInstance.AddModuleMessage("message", MessageType.Info); to AddModuleMessage("message", MessageType.Info); (ie. remove the ModuleInstance reference)

Unfortunately, the usage of ModuleInstance.AddModuleMessage("message", MessageType.Info); is fairly common, as it was included in the default module template (in the Settings component). So in many cases developers will need to update their modules to ensure functionality behaves as intended.

**Comment by sbwalker** [comment]
One other change to be aware of is that when running a Site on Static render mode, the Theme and Container components are rendered statically. This means that they cannot contain  interactive functionality such as ```@onlick()``` event handlers or use JS Interop. If your theme or containers contain interactive functionality you will need to run your Site in Interactive render mode (which enables interactivity in ALL components).

**Comment by sbwalker** [comment]
I am curious if the configuration options would make more sense in the Site Settings Ui if they were:

UI Component Settings

Render Mode - Interactive / Static
Interactivity - Server / Client / Auto

This would make them less "technical" which may be good for end users (but would maybe cause confusion for developers who seem to like the very technical jargon such as SignalR, WebAssembly, etc...)

Your thoughts are welcome.

**Reply by mdmontesinos** [reply]
I believe the current naming should be fine as it is clearer for developers and aligns with Blazor conventions. Does it really make sense for an end user to configure render modes and interactivity? Does an end user know the implications of each mode?

**Reply by thabaum** [reply]
```
Render Mode - Interactive / Static
Interactivity - Server / Client / Auto
```

Personally I like this, if not `WebAssembly` instead of `Client`, however from an administration perspective `Client` should work best as understood by someone that maybe is not a Blazor developer, on the other hand a programmer would probably prefer `WebAssembly` but they would understand the difference.

**Comment by iJungleboy** [comment]
@sbwalker this is impressive as 🔥!!!

Very, very cool. 

Can't wait to try it.

One idea - which is something I had to introduce in 2sxc as time progressed:

How about making a new base class like `ModuleBase5` which would have the new defaults? This strategy has let me create new base classes and remove old / deprecated features.

My strategy is also to only add new features to the latest base class, to encourage users to drop old habits.

**Reply by sbwalker** [reply]
@iJungleboy introducing a new base class would not resolve the problem described above. There are no new "defaults". The AddModuleMessage() and ShowProgressIndicator() methods that are part of the base class are still relevant and widely used. They are just implemented differently under the covers so that they can function properly. The only problem is the reference to ModuleInstance which is part of the base class - which was a legacy approach to accessing the AddModuleMessage() and ShowProgressIndicator() methods which was later superseded with native methods in ModuleBase.

**Reply by iJungleboy** [reply]
np. I was just referring to the `RenderMode` being an _old_ default, and a new base class could provide the _new_ default `Static`

**Reply by sbwalker** [reply]
OK I understand what you mean now. I actually believe that for many scenarios, the Interactive method for developing modules is the best approach (ie. any administrative user interface with high interactivity) - so the existing ModuleBase will not be phased out or deprecated. There may however be a benefit in also creating a new StaticModuleBase which inherits from ModuleBase and simply overrides the RenderMode default value. This would eliminate the need to set RenderMode in each static module component for static modules.

**Comment by sbwalker** [comment]
I am also still considering whether the default render mode for new Sites should be Static Server Rendering. This would align with the expectation of developers who are familiar with .NET 8 as SSR is the new default app template for Blazor in Visual Studio.

**Reply by DaveAndersonCo** [reply]
To me it also makes more sense to follow the new default app template for Blazor.
